### ISSUE TITLE
Something wrong in your decoding functions

### ISSUE THREAD

**Issue by YuryStrozhevsky** [emailjs/emailjs-mime-parser#24]
I was trying to decode correctly binary data using latest version of `emailjs-mime-parser` but failed. In fact I was using it in [this](https://github.com/PeculiarVentures/PKI.js/tree/new_smime_examples/examples/SMIMEEncryptionExample) and [this](https://github.com/PeculiarVentures/PKI.js/tree/new_smime_examples/examples/SMIMEVerificationExample) examples.

So, I made a simple `unit test` file for your package. Here it is:
```javascript
import parse from './mimeparser';
//**************************************************************************************
function bufferToHexCodes(inputBuffer, inputOffset = 0, inputLength = inputBuffer.byteLength)
{
	let result = "";
	
	for(const item of (new Uint8Array(inputBuffer, inputOffset, inputLength)))
	{
		const str = item.toString(16).toUpperCase();
		result = result + ((str.length === 1) ? "0" : "") + str;
	}
	
	return result;
}

//**************************************************************************************
context("Test", () =>
{
	it("test 1", () =>
	{
		const testMessage = `Content-Type: application/pkcs7-mime; name=smime.p7m; smime-type=enveloped-data; charset=binary
Content-Description: Enveloped Data
Content-Disposition: attachment; filename=smime.p7m
Content-Transfer-Encoding: base64
From: sender@example.com
To: recipient@example.com
Subject: Example S/MIME encrypted message
Date: Sun, 25 Feb 2018 09:28:14 +0000
Message-Id: <1519550894482-b208bdc8-7f8e90aa-4af6b4fa@example.com>
MIME-Version: 1.0

MIIB3AYJKoZIhvcNAQcDoIIBzTCCAckCAQIxggFuMIIBagIBADAjMB4xHDAJBgNVBAYTAlJVMA8G
A1UEAx4IAFQAZQBzAHQCAQEwPAYJKoZIhvcNAQEHMC+gDzANBglghkgBZQMEAgMFAKEcMBoGCSqG
SIb3DQEBCDANBglghkgBZQMEAgMFAASCAQBDyepahKyM+hceeF7J+pyiSVYLElKyFKff9flMs1VX
ZaBQRcEYpIqw9agD4u+aHlIOJ6AtdCbxaV0M8q6gjM4E5lUFUOqG/QIycdG2asZ0lza/DL8SdxfA
3WE9Ij5IEqFbtnykbfORK+5XWT0nYs/OMN0NKeCwXjElNsezX9IAIgxHgwcVYW+szXpRlarjriAC
TDG/M+Xl5YtyAhmHWFncBSfWM8e2q+AKh3eCal1lH4eXtGICc4rad4f6845YJwXL8DYYS+GdLVAY
EXKuHr0N7g4aHTs9B8EQqHmYdaHWTi3h0ZPkvAE+wwfm9xjvL2z2HrfpYyMTvALrefvSt7sGMIAG
CSqGSIb3DQEHATAdBglghkgBZQMEAQIEEKt6VqFcNz/VYFwu85DTOqGggAQgIHc45LBiYIQqhxNw
hlRk4BxMiyiQRdLcVdCwwkKyX2sAAAAA`;
		
		const expectedText = "308201DC06092A864886F70D010703A08201CD308201C90201023182016E3082016A0201003023301E311C3009060355040613025255300F06035504031E080054006500730074020101303C06092A864886F70D010107302FA00F300D06096086480165030402030500A11C301A06092A864886F70D010108300D060960864801650304020305000482010043C9EA5A84AC8CFA171E785EC9FA9CA249560B1252B214A7DFF5F94CB3555765A05045C118A48AB0F5A803E2EF9A1E520E27A02D7426F1695D0CF2AEA08CCE04E6550550EA86FD023271D1B66AC6749736BF0CBF127717C0DD613D223E4812A15BB67CA46DF3912BEE57593D2762CFCE30DD0D29E0B05E312536C7B35FD200220C47830715616FACCD7A5195AAE3AE20024C31BF33E5E5E58B720219875859DC0527D633C7B6ABE00A8777826A5D651F8797B46202738ADA7787FAF38E582705CBF036184BE19D2D50181172AE1EBD0DEE0E1A1D3B3D07C110A8799875A1D64E2DE1D193E4BC013EC307E6F718EF2F6CF61EB7E9632313BC02EB79FBD2B7BB06308006092A864886F70D010701301D06096086480165030401020410AB7A56A15C373FD5605C2EF390D33AA1A0800420207738E4B06260842A871370865464E01C4C8B289045D2DC55D0B0C242B25F6B00000000";
		
		const root = parse(testMessage);
		
		expect(bufferToHexCodes(root.content.buffer)).to.equal(expectedText);
	});
});
//**************************************************************************************
```

The `expectedText` value I got by decoding BASE-64 text from message by my own utility. You could check that `expectedText` value is 100% correct. Here are indexes (indexes of `expectedText` string) where we have differences:
```
355, 362, 363, 479, 636, 637, 692, 693, 714, 715, 738, 739, 774, 775
``` 

Please correct the issue or provide me advice how to fix that.

**Comment by YuryStrozhevsky** [comment]
Improved `unit test` code:
```javascript
import parse from './mimeparser';
//**************************************************************************************
context("Test", () =>
{
	it("test 1", () =>
	{
		const testHeader = `Content-Type: application/pkcs7-mime; name=smime.p7m; smime-type=enveloped-data; charset=binary
Content-Description: Enveloped Data
Content-Disposition: attachment; filename=smime.p7m
Content-Transfer-Encoding: base64
From: sender@example.com
To: recipient@example.com
Subject: Example S/MIME encrypted message
Date: Sun, 25 Feb 2018 09:28:14 +0000
Message-Id: <1519550894482-b208bdc8-7f8e90aa-4af6b4fa@example.com>
MIME-Version: 1.0

`;
		
		const testMessage = `MIIB3AYJKoZIhvcNAQcDoIIBzTCCAckCAQIxggFuMIIBagIBADAjMB4xHDAJBgNVBAYTAlJVMA8G
A1UEAx4IAFQAZQBzAHQCAQEwPAYJKoZIhvcNAQEHMC+gDzANBglghkgBZQMEAgMFAKEcMBoGCSqG
SIb3DQEBCDANBglghkgBZQMEAgMFAASCAQBDyepahKyM+hceeF7J+pyiSVYLElKyFKff9flMs1VX
ZaBQRcEYpIqw9agD4u+aHlIOJ6AtdCbxaV0M8q6gjM4E5lUFUOqG/QIycdG2asZ0lza/DL8SdxfA
3WE9Ij5IEqFbtnykbfORK+5XWT0nYs/OMN0NKeCwXjElNsezX9IAIgxHgwcVYW+szXpRlarjriAC
TDG/M+Xl5YtyAhmHWFncBSfWM8e2q+AKh3eCal1lH4eXtGICc4rad4f6845YJwXL8DYYS+GdLVAY
EXKuHr0N7g4aHTs9B8EQqHmYdaHWTi3h0ZPkvAE+wwfm9xjvL2z2HrfpYyMTvALrefvSt7sGMIAG
CSqGSIb3DQEHATAdBglghkgBZQMEAQIEEKt6VqFcNz/VYFwu85DTOqGggAQgIHc45LBiYIQqhxNw
hlRk4BxMiyiQRdLcVdCwwkKyX2sAAAAA`;
		
		const expectedText = (Buffer.from(testMessage, "base64")).toString("hex").toUpperCase();
		
		const root = parse(testHeader + testMessage);
		
		expect((Buffer.from(root.content.buffer)).toString("hex").toUpperCase()).to.equal(expectedText);
	});
});
//**************************************************************************************
```

**Comment by felixhammerl** [comment]
Thanks for reporting this. The problem seems to be rather in the base64 encoding part than in the MIME parsing.

**Comment by felixhammerl** [comment]
Ok, that doesn't seem to be it. Added test case for b64 parsing here: https://github.com/emailjs/emailjs-base64/commit/d4677a7c8d3a2a0768eaced45945c341e19f6e55

**Comment by YuryStrozhevsky** [comment]
As I understood the problem is that during data initializing (processLine routine) you applying a set of transformation, not only decoding from base64. So your base64 has correct transformation but after that you apply utf8 transformation and that is why have broken output data.

**Comment by felixhammerl** [comment]
Fixed in v2.0.3
